packages/yarnpkg-core/sources/scriptUtils.ts

```diff
@@ -376,7 +376,7 @@ async function initializeWorkspaceEnvironment(workspace: Workspace, {binFolder,
     )
   );
 
-  return {manifest: workspace.manifest, binFolder, env, cwd: cwd ?? workspace.cwd};
+  return {manifest: workspace.manifest, binFolder, env, cwd: cwd ?? ppath.dirname(await xfs.realpathPromise(ppath.join(workspace.cwd, `package.json` as Filename)))};
```

Comment:
Why the dirname/join? Realpath should return the real path even if the last component is a folder, isn't it?

Reply:
This mimicks the behavior before the optimization change. Also for projects using bazel, which is probably the most common situation where you would run into symlinks, (and also our use case) folders are not linked but files are.